### PR TITLE
I/O Bindings for ev

### PR DESCRIPTION
**Pull request by waht**
# Purpose

I/O Bindings integration for [libev](http://software.schmorp.de/pkg/libev.html).

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I added unit tests
- [x] I ran all tests and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] release notes are updated (doc/news/_preparation_next_release.md)

This PR is based on #1833. I will rebase on `master` after it is merged.

@markus2330 please review my pull request

### CODE REVIEW

**Comment by waht** [comment]
jenkins build all please

File: cmake/Modules/FindLibEv.cmake [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# EOS - The CERN Disk Storage System
+# Copyright (C) 2015 CERN/Switzerland
+# This program is free software: you can redistribute it and/or modify it under
```
Comment (markus2330) [inline cmake/Modules/FindLibEv.cmake]: Could you please ask them if they can relicense this trivial code as BSD/public domain? If it is copyright-able at all: The license header is as long as the code. Where did you get it from?

File: src/bindings/README.md [hunk 2]
```diff
@@ -25,6 +25,8 @@ Experimental bindings (included in `EXPERIMENTAL`):
 - [haskell](haskell/) Haskell bindings
 - [intercept_fs](intercept/fs/) Intercepts file system calls to configuration files (experimental)
 - [io_uv](io/uv/) I/O bindings for uv (experimental)
+- [io_glib](io/glib/) I/O bindings for glib (experimental)
+- [io_ev](io/ev/) I/O bindings for ev (experimental)
```
Comment (markus2330) [inline src/bindings/README.md]: Maybe group uv+ev next to each other as they are historically related.

File: src/bindings/io/CMakeLists.txt [hunk 3]
```diff
@@ -3,6 +3,16 @@ if (IS_INCLUDED)
 	add_subdirectory (uv)
 endif ()
 
+check_binding_included ("io_ev" IS_INCLUDED SUBDIRECTORY "io/ev")
```
Comment (markus2330) [inline src/bindings/io/CMakeLists.txt]: The order here is already correct :-)

File: src/bindings/io/ev/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+Make sure to build your application with `elektra-io-ev`, `elektra-io` and `ev` or
+simply use `pkg-config --cflags --libs elektra-io-ev`.
+
+### ElektraIoInterface * elektraIoEvNew (struct ev_loop * loop)
+
+Create and initialize a new I/O binding.
+
+*Parameters*
+
+- loop Loop to use for I/O operations
```
Comment (markus2330) [inline src/bindings/io/ev/README.md]: loop Loop, either remove one word or use ":"

File: src/bindings/io/glib/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+Make sure to build your application with `elektra-io-glib`, `elektra-io` and `glib` or
+simply use `pkg-config --cflags --libs elektra-io-glib`.
+
+### ElektraIoInterface * elektraIoGlibNew (GMainContext * context)
+
+Create and initialize a new I/O binding.
+
+*Parameters*
+
+- context Context to use for I/O operations. May be NULL to indicate the glib's
```
Comment (markus2330) [inline src/bindings/io/glib/README.md]: again

File: src/bindings/io/glib/testio_glib.c [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <kdbio.h>
+#include <kdbiotest.h>
+#include <tests.h>
+
+#include <glib.h>
+
+#include "kdbio_glib.h"
+
+#include <unistd.h> // pipe()
```
Comment (markus2330) [inline src/bindings/io/glib/testio_glib.c]: Is it really used?

File: cmake/Modules/FindLibEv.cmake (hunk 1 above)
Comment (waht) [reply cmake/Modules/FindLibEv.cmake]: There are a few available online. I found some GPLv2, GPLv3, Apache, MIT, but no BSD. The [original file](https://gitlab.cern.ch/dss/eos/blob/ea596793395a846038e5e02326d804a6b9cda8d8/cmake/Findlibev.cmake) had no license header, I added the project's license header.

**Comment by markus2330** [comment]
jenkins build all please

**Comment by markus2330** [comment]
Seems like reformatting is necessary.

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
Yeah, I didn't notice we changed to clang-format 5.0 :)

**Comment by markus2330** [comment]
jenkins build all please

**Comment by markus2330** [comment]
sorry, I needed to restart jenkins due to high load

> Yeah, I didn't notice we changed to clang-format 5.0 :)

Actually we now support clang-format 5-7 but `.clang-format` needed changes in order to support his.

**Comment by waht** [comment]
jenkins build doc please

**Comment by waht** [comment]
jenkins build gcc-i386 please

**Comment by waht** [comment]
jenkins build test-bindings please

**Comment by waht** [comment]
jenkins build homepage please

**Comment by waht** [comment]
jenkins build incremental please

**Comment by waht** [comment]
jenkins build incremental please

**Comment by haraldg** [comment]
Unfortunately for me the busy season has started and I have little time for a real code review (or any programming actually). However I had a quick look and I feel I should point out that I don't undestand at all what problems these ElektraIo bindings try to solve:

AFAICS this is just a wrapper around libev? - Why would I want that? Why would elektra manage my file descriptors?

I was expecting something that allows to watch for configuration related events from an event loop, but the example given doesn't relate to configuration at all. I guess all this makes sense, if the binding is only used internally by elektra. However then I don't see what the example actually demonstrates ...

A few random comments:
* It is good practise to use only non-blocking file descriptors with libev. Since I don't have the big picture a bit difficult for me to tell if this is relevant here.
* AFAIK memory allocation doesn't fail on any operating systems nowadays: You sooner get a signal than an invalid pointer. In my own code I either treat it as non-failing or (if following the standards to the letter is important) crash the application on failure. I think any API that carries error conditions around only because of memory allocation, is needlessly complex.

@markus2330 is there anything in particular, where you would like the get my input?

**Comment by markus2330** [comment]
Thank you, you already gave good feedback: The high-level view (use case) is completely missing in the documentation and the example seems to be not as self-describing as we thought it would be.

**Comment by waht** [comment]
Thank you for your feedback @haraldg!

I/O bindings allow Elektra's plugins to integrate into different main loop APIs using a thin binding layer. For example, this is used for notification transport plugins which receive notifications using ZeroMQ/D-Bus/etc. In such cases a socket file descriptor to watch for available messages is used - those file descriptors should be non-blocking.

The input on memory allocation is helpful. I will check the notification API for this. For this binding memory allocation is the only way some functions can fail, but other APIs like glib or uv may have other error conditions.

**Comment by markus2330** [comment]
> I/O bindings allow [...]

This should be stated clearly in the entry points of the docu. It is not so obvious what your libs do.

> AFAIK memory allocation doesn't fail on any operating systems nowadays

If you allocate very large junks you can easily provoke malloc to fail. But yes, you usually get much more than actually is available. But also this behavior can be configured, so there is no guarantee in any direction.

**Comment by haraldg** [comment]
JFTR: It wasn't the docu that got me confused, but this example programm that adds a watcher for stdin.

> If you allocate very large junks you can easily provoke malloc to fail.

Sure, but in cases where the user code can't influence the size of memory allocated, this is often better checked by an assert() than returning an error condition.

> But also this behavior can be configured, so there is no guarantee in any direction.

Of course there is no guarantee. Since the standards allow returning an error condition any libc is free to do so. My point is: On any systems (using virtual memory) that I know of, at the point where malloc fails (if at all) the system is in such a bad shape, that recovering is pretty much impossible. Or to put it in other words: IMO a library function should only return an error condition, if the user code can actually do something about it. Delaying a crash usually just makes everything worse.

I don't know of a portable way to configure this behaviour. On linux you would need to disable overcommitting and also disable swapping - but actually swapping isn't disabled when you have zero swapspace, the kernel is still free to discard any mmaped pages.

There is also setrlimit(RLIMIT_DATA, ...) that is fairly portable and might cause malloc() to fail, but again I don't quite see what a process should do, when it has a too tight memory limit.

**Comment by markus2330** [comment]
> Sure, but in cases where the user code can't influence the size of memory allocated, this is often better checked by an assert() than returning an error condition.

Yes, elektraMalloc does exactly that. Unfortunately, `assert`s are usually turned off in productive builds... You can turn it on with `ELEKTRA_DEBUG` (with any optimization level set by `CMAKE_BUILD_TYPE`).

> On any systems (using virtual memory) that I know of, at the point where malloc fails (if at all) the system is in such a bad shape, that recovering is pretty much impossible.
> [...]
> but again I don't quite see what a process should do, when it has a too tight memory limit.

Good to hear that also from someone with embedded background. It takes away a lot of pressure :+1: 

Nevertheless, even if the implementation does not check for the return value of malloc, someone else might reimplement the APIs with much higher reliability in mind. But in any case we need to look at individual APIs, with general statements we won't get far.

**Comment by haraldg** [comment]
> Unfortunately, asserts are usually turned off in productive builds...

Not unfortunate at all. This is why I suggested them in the first place.

> Good to hear that also from someone with embedded background. It takes away a lot of pressure

Actually 'embedded background' is one of the reasons why I even think about such questions: Often unnecessary error handling takes more space then the data structure being allocated ... 

> Nevertheless, even if the implementation does not check for the return value of malloc, someone else might reimplement the APIs with much higher reliability in mind.

I still don't believe this is a way to improve reliability - or rather: I believe that any process (other then init) that tries to keep running as much as possible, hurts the reliability of the overall system. But I agree with your distinction between API and implementation: If you want the elektra API to be portable to platforms without virtual memory, then this imposes additinal requirements.

> But in any case we need to look at individual APIs, with general statements we won't get far.

Jein. ;)

On one hand I think that the core question (what should/could user code do, when a malloc of a harmless data structure fails?) is quite general, on the other hand there are some corner cases (like freeing resources, that won't be reclaimed by the operating systems, when the application crashes) that need special consideration.

Of course this is just the technical side. I can see that it might politically impractical to convince all potential users of libelektra, that sometimes crashing their application is the right thing.

Anyway: After thinking about your API a bit, I came up with a different question: Do you intend to support using multiple I/O bindings at the same time? If not, then maybe matters can be simplified a lot, by just implementing the same interface (same symbols) in each binding instead of passing pointers to data structures around.

**Comment by haraldg** [comment]
I just realized that I did a bit of a poor job at explaining the rationale behind my previous comments because I failed to mention something that's probably not common wisdom: To have an API support reliable programming on the level of "the system might run out of memory" the user code needs to allocate/provide the memory for all required data structures. - Then the user code can mlock() the memory region to ensure, that it always stays in physical ram.

**Comment by markus2330** [comment]
I moved the memory management discussion to #1873

> Do you intend to support using multiple I/O bindings at the same time?

Yes, this should be clearly answered in some high-level architectural description.

**Comment by markus2330** [comment]
> I was expecting something that allows to watch for configuration related events from an event loop, but the example given doesn't relate to configuration at all. I guess all this makes sense, if the binding is only used internally by elektra. However then I don't see what the example actually demonstrates ...

Yes, also the text in the comment above is hard to understand or misleading. It is hard to understand "In reality code using the I/O binding will attach non-blocking file descriptors (e.g. from sockets)" if it was unclear what the example is about at all.

The other examples (glib, uv) have similar problems. It will already help a lot if you point to "./src/libs/notification/example" which is more likely what the user is actually interested in.

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
I added an [overview for I/O bindings](https://github.com/ElektraInitiative/libelektra/pull/1840/files#diff-c48d19dc33fe53b714f9180132e5f06aR20) to the developer API and I/O binding examples [now link](https://github.com/ElektraInitiative/libelektra/pull/1840/files#diff-3fb79e9cc4588a355f603efe215ffc76R8) to the notification example.

I think the PR is ready to merge.

**Comment by markus2330** [comment]
In https://build.libelektra.org/jenkins/job/elektra-clang-asan/427/console testlib_notification leaks/fails.

**Comment by waht** [comment]
jenkins build all please

File: src/bindings/io/ev/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	elektraIoSetBinding (kdb, binding);
+
+	// Start the event loop
+	ev_run (loop, 0);
+
+	// Cleanup before exit
+	elektraIoBindingCleanup (binding);
+	ev_loop_destroy (EV_DEFAULT);
+}
+
```
[the remark this answers is not in the document]
Comment (markus2330) [reply src/bindings/io/ev/README.md]: Any reason why there is newline at beginning and end? Is there some formatter where this looks better?

File: src/bindings/io/ev/elektra-io-ev.pc.in [hunk 8]
```diff
@@ -0,0 +1,14 @@
+prefix=@CMAKE_INSTALL_PREFIX@
```
Comment (markus2330) [inline src/bindings/io/ev/elektra-io-ev.pc.in]: Thank you for writing these files.

File: src/bindings/io/ev/io_ev.c [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+/**
+ * Add idle operation to I/O binding
+ * @see kdbio.h ::ElektraIoBindingAddIdle
+ */
+static int ioEvBindingAddIdle (ElektraIoInterface * binding ELEKTRA_UNUSED, ElektraIoIdleOperation * idleOp)
+{
+	EvBindingData * bindingData = newBindingData ();
+	if (bindingData == NULL)
+	{
+		return 0;
```
[the remark this answers is not in the document]
Comment (markus2330) [reply src/bindings/io/ev/io_ev.c]: Could you elaborate here, please?

**Comment by markus2330** [comment]
It looks okay now, I only reasked 2 questions.

Please set "ready to merge" if ready to merge.

**Comment by haraldg** [comment]
Sorry, no time ATM. Also unless there was an API change since I last looked, there probably isn't any questions that I care about anyway.

File: src/bindings/io/ev/io_ev.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (waht) [reply src/bindings/io/ev/io_ev.c]: Ah I see, I must have missed the first comment the last time.

The difference is that malloc() in newBindingData() might fail while the assertions are only triggered on API misuse.
Both functions are user-facing so I'll turn the assertions into explicit checks.

File: src/bindings/README.md [hunk 10]
```diff
@@ -25,6 +25,7 @@ Experimental bindings (included in `EXPERIMENTAL`):
 - [haskell](haskell/) Haskell bindings
 - [intercept_fs](intercept/fs/) Intercepts file system calls to configuration files (experimental)
 - [io_uv](io/uv/) I/O bindings for uv (experimental)
```
[the remark this answers is not in the document]
Comment (waht) [reply src/bindings/README.md]: Good point! Currently the uv binding is used for testing different plugins internally. The ev binding was used in the  libautomation benchmarks.

**Comment by waht** [comment]
jenkins build all please

**Comment by waht** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**